### PR TITLE
Update collection-expressions.md

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Additional tweaks and clarifications.

### CODE REVIEW

File: proposals/csharp-12.0/collection-expressions.md [hunk 1]
```diff
@@ -715,16 +735,21 @@ However, given the breadth and consistency brought by the new literal syntax, we
...
+
+* Should we expand on collection initializers to look for the very common `AddRange` method? It could be used by the underlying constructed type to perform adding of spread elements potentially more efficiently.  We might also want to look for things like `.CopyTo` as well.  There may be drawbacks here as those methods might end up causing excess allocations/dispatches versus directly enumerating in the translated code.
+
+    Yes.  An implementation is allowed to utilize other methods to initialize a collection value, under the presumption that these methods have well-defined semantics, and that collection types should be "well behaved".  In practice though, an implementation should be cautious as benefits in one way (bulk copying) may come with negative consequences as well (for example, boxing a struct collection).
+
+    An implementation should take advantage in the cases where there are no downsides (like no boxing).  For example, with a `.AddRange(ReadOnlySpan<T>)` method.
```
Comment (CyrusNajmabadi) [inline proposals/csharp-12.0/collection-expressions.md:746]:
```suggestion
    An implementation should take advantage in the cases where there are no downsides.  For example, with a `.AddRange(ReadOnlySpan<T>)` method.
```

File: proposals/csharp-12.0/collection-expressions.md [hunk 2]
```diff
@@ -98,28 +100,23 @@ Collection literals are [target-typed](https://github.com/dotnet/csharplang/blob
...
 [conversions]: #conversions
 
-A *collection literal conversion* allows a collection literal expression to be converted to a type.
+A *collection expression conversion* allows a collection literal expression to be converted to a type.
 
-The following implicit *collection literal conversions* exist from a collection literal expression:
+The following implicit *collection expression conversions* exist from a collection literal expression:
 
 * To a single dimensional *array type* `T[]` where:
-
```
Comment (cston) [inline proposals/csharp-12.0/collection-expressions.md:106]: The blank lines were added in https://github.com/dotnet/csharplang/pull/7468, and might be necessary.

Comment (CyrusNajmabadi) [reply proposals/csharp-12.0/collection-expressions.md:106]: ok :)

File: proposals/csharp-12.0/collection-expressions.md [hunk 3]
```diff
@@ -98,28 +100,23 @@ Collection literals are [target-typed](https://github.com/dotnet/csharplang/blob
 ## Conversions
 [conversions]: #conversions
 
-A *collection literal conversion* allows a collection literal expression to be converted to a type.
+A *collection expression conversion* allows a collection literal expression to be converted to a type.
 
-The following implicit *collection literal conversions* exist from a collection literal expression:
+The following implicit *collection expression conversions* exist from a collection literal expression:
 
 * To a single dimensional *array type* `T[]` where:
```
Comment (CyrusNajmabadi) [inline proposals/csharp-12.0/collection-expressions.md:107]:
```suggestion
* To a single dimensional *array type* `T[]` where:

```

File: proposals/csharp-12.0/collection-expressions.md [hunk 4]
```diff
@@ -98,28 +100,23 @@ Collection literals are [target-typed](https://github.com/dotnet/csharplang/blob
...
-
   * For each *element* `Ei` there is an *implicit conversion* to `T`.
 
 * To a *span type* `System.Span<T>` or `System.ReadOnlySpan<T>` where:
```
Comment (CyrusNajmabadi) [inline proposals/csharp-12.0/collection-expressions.md:110]:
```suggestion
* To a *span type* `System.Span<T>` or `System.ReadOnlySpan<T>` where:

```

File: proposals/csharp-12.0/collection-expressions.md [hunk 5]
```diff
@@ -98,28 +100,23 @@ Collection literals are [target-typed](https://github.com/dotnet/csharplang/blob
...
-
   * For each *element* `Ei` there is an *implicit conversion* to `T`.
 
 * To a *type* with a *[create method](#create-methods)* with *parameter type* `System.ReadOnlySpan<T>` where:
```
Comment (CyrusNajmabadi) [inline proposals/csharp-12.0/collection-expressions.md:113]:
```suggestion
* To a *type* with a *[create method](#create-methods)* with *parameter type* `System.ReadOnlySpan<T>` where:

```

File: proposals/csharp-12.0/collection-expressions.md [hunk 6]
```diff
@@ -98,28 +100,23 @@ Collection literals are [target-typed](https://github.com/dotnet/csharplang/blob
...
-
   * For each *element* `Ei` there is an *implicit conversion* to `T`.
 
 * To an *[inline array type](https://github.com/dotnet/csharplang/blob/main/proposals/inline-arrays.md)* with *element type* `T` where:
```
Comment (CyrusNajmabadi) [inline proposals/csharp-12.0/collection-expressions.md:116]:
```suggestion
* To an *[inline array type](https://github.com/dotnet/csharplang/blob/main/proposals/inline-arrays.md)* with *element type* `T` where:

```

File: proposals/csharp-12.0/collection-expressions.md [hunk 7]
```diff
@@ -489,7 +505,9 @@ Not having a *known length* does not prevent any result from being created. Howe
 
   * If `T` is some `ReadOnlySpan<T1>`, then the literal is translated the same as for the `Span<T1>` case except that the final result will be that `Span<T1>` [implicitly converted](https://learn.microsoft.com/dotnet/api/system.span-1.op_implicit#system-span-1-op-implicit(system-span((-0)))-system-readonlyspan((-0))) to a `ReadOnlySpan<T1>`.
 
-    The above forms (for arrays and spans) are the base representations of the literal value and are used for the following translation rules.
+    A `ReadOnlySpan<T1>` where `T1` is some primitive type, and all collection elements are constant does not need its data to be on the heap, or on the stack.  For example, an implementation could construct this span directly as a reference to portion of the data segment of the program.
+
+    The above forms (for arrays and spans) are the base representations of the literal value and are used for the following translation rules:
```
Comment (cston) [inline proposals/csharp-12.0/collection-expressions.md:515]:
```suggestion
    The above forms (for arrays and spans) are the base representations of the collection expression and are used for the following translation rules:
```

File: proposals/csharp-12.0/collection-expressions.md [hunk 8]
```diff
@@ -715,16 +740,21 @@ However, given the breadth and consistency brought by the new literal syntax, we
...
+* Should we expand on collection initializers to look for the very common `AddRange` method? It could be used by the underlying constructed type to perform adding of spread elements potentially more efficiently.  We might also want to look for things like `.CopyTo` as well.  There may be drawbacks here as those methods might end up causing excess allocations/dispatches versus directly enumerating in the translated code.
+
+    Yes.  An implementation is allowed to utilize other methods to initialize a collection value, under the presumption that these methods have well-defined semantics, and that collection types should be "well behaved".  In practice though, an implementation should be cautious as benefits in one way (bulk copying) may come with negative consequences as well (for example, boxing a struct collection).
+
+    An implementation should take advantage in the cases where there are no downsides.  For example, with a `.AddRange(ReadOnlySpan<T>)` method.
```
Comment (CyrusNajmabadi) [inline proposals/csharp-12.0/collection-expressions.md:751]:
```suggestion
    An implementation should take advantage in the cases where there are no downsides.  For example, with an `.AddRange(ReadOnlySpan<T>)` method.
```

File: proposals/csharp-12.0/collection-expressions.md [hunk 9]
```diff
@@ -715,16 +735,21 @@ However, given the breadth and consistency brought by the new literal syntax, we
...
+    Yes. The compiler should emit `Array.Empty<T>()` for any case where this is legal and the final result is non-mutable.  For example, targeting `T[]`, `IEnumerable<T>`, `IReadOnlyCollection<T>` or `IReadOnlyList<T>`.  It should not use `Array.Empty<T>` when the target is mutable (`ICollection<T>` or `IList<T>`).
+
+* Should we expand on collection initializers to look for the very common `AddRange` method? It could be used by the underlying constructed type to perform adding of spread elements potentially more efficiently.  We might also want to look for things like `.CopyTo` as well.  There may be drawbacks here as those methods might end up causing excess allocations/dispatches versus directly enumerating in the translated code.
+
+    Yes.  An implementation is allowed to utilize other methods to initialize a collection value, under the presumption that these methods have well-defined semantics, and that collection types should be "well behaved".  In practice though, an implementation should be cautious as benefits in one way (bulk copying) may come with negative consequences as well (for example, boxing a struct collection).
```
Comment (cston) [inline proposals/csharp-12.0/collection-expressions.md:749]: I think the question of which methods the compiler can and should use is still an open question. And we need to specific with the set of methods that might be used.

Comment (CyrusNajmabadi) [reply proposals/csharp-12.0/collection-expressions.md:749]: Can we have the spec say that this is allowed, but have it be a compiler investigation as to what we actually use?

File: proposals/csharp-12.0/collection-expressions.md [hunk 10]
```diff
@@ -98,9 +100,9 @@ Collection literals are [target-typed](https://github.com/dotnet/csharplang/blob
 ## Conversions
 [conversions]: #conversions
 
-A *collection literal conversion* allows a collection literal expression to be converted to a type.
+A *collection expression conversion* allows a collection literal expression to be converted to a type.
```
Comment (333fred) [inline proposals/csharp-12.0/collection-expressions.md:103]:
```suggestion
A *collection expression conversion* allows a collection expression to be converted to a type.
```

File: proposals/csharp-12.0/collection-expressions.md [hunk 11]
```diff
@@ -98,9 +100,9 @@ Collection literals are [target-typed](https://github.com/dotnet/csharplang/blob
...
 
-The following implicit *collection literal conversions* exist from a collection literal expression:
+The following implicit *collection expression conversions* exist from a collection literal expression:
```
Comment (333fred) [inline proposals/csharp-12.0/collection-expressions.md:105]:
```suggestion
The following implicit *collection expression conversions* exist from a collection expression:
```

File: proposals/csharp-12.0/collection-expressions.md [hunk 12]
```diff
@@ -70,7 +70,7 @@ Collection literals are [target-typed](https://github.com/dotnet/csharplang/blob
 ### Spec clarifications
 [spec-clarifications]: #spec-clarifications
 
-* For brevity, `collection_literal_expression` will be referred to as "literal" in the following sections.
+* For brevity, `collection_expression` will be referred to as "literal" in the following sections.
```
Comment (333fred) [inline proposals/csharp-12.0/collection-expressions.md:73]: Probably would reword this to remove references to `literal`

File: proposals/csharp-12.0/collection-expressions.md [hunk 13]
```diff
@@ -229,6 +231,21 @@ ImmutableArray<int> ia =
 
     Here, if `b` is false, it is not required that any value actually be constructed for the empty literal since it would immediately be spread into zero values in the final literal.
```
Comment (333fred) [inline proposals/csharp-12.0/collection-expressions.md:232]:
```suggestion
    Here, if `b` is false, it is not required that any value actually be constructed for the empty collection expression since it would immediately be spread into zero values in the final literal.
```

File: proposals/csharp-12.0/collection-expressions.md [hunk 14]
```diff
@@ -229,6 +231,21 @@ ImmutableArray<int> ia =
...
 
+* The empty literal is permitted to be a singleton if used to construct a final collection value that is known to not be mutable.  For example:
```
Comment (333fred) [inline proposals/csharp-12.0/collection-expressions.md:234]:
```suggestion
* The empty collection expression is permitted to be a singleton if used to construct a final collection value that is known to not be mutable.  For example:
```

**Review by 333fred — COMMENTED** [review]
If we're going to replace `literal` in the spec, it should be more thorough. I've pointed out a few places, but there are certainly more.

**Comment by CyrusNajmabadi** [comment]
Merging in now.  Will do another pass on 'literal', and will flesh out section on allowed apis the compiler can call.
